File: dotnet/Models/Graph.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;

namespace tube.Models
{
    public class Graph
    {
        private List<Node> _nodes;

        public Graph()
        {
            _nodes = new List<Node>();
        }

        public void AddEdge(string fromStation, string toStation)
        {
            var first = GetNodeByName(fromStation);
            var second = GetNodeByName(toStation);

            if (first == null)
            {
                first = new Node { Name = fromStation };
                _nodes.Add(first);
            }

            if (second == null) 
            {
                second = new Node { Name = toStation };
                _nodes.Add(second);
            }

            first.AddEdge(second);
            second.AddEdge(first);
        }

        public Node GetNodeByName(string name)
        {
            return _nodes.FirstOrDefault(s => s.Name == name);
        }
    }
}

File: kotlin/src/Models/Graph.kt
package Models

class Graph {
    private var _nodes: MutableList<Node> = mutableListOf()

    fun addEdge(fromStation: String, toStation: String) {
        var first = getNodeByName(fromStation)
        var second = getNodeByName(toStation)

        if (first == null) {
            first = Node(fromStation)
            _nodes.add(first)
        }

        if (second == null) {
            second = Node(toStation)
            _nodes.add(second)
        }

        first.addEdge(second)
        second.addEdge(first)
    }

    fun getNodeByName(name: String?): Node? {
        if (!name.isNullOrEmpty()) {
            return _nodes.find { node -> node.name == name }
        }
        return null
    }
}
